packages/mdl-textfield/foundation.js

```diff
@@ -0,0 +1,153 @@
+      rmHelptextAttr: (/* name: string, value: string */) => {},
+      getNativeInput: () => /* HTMLInputElement */ ({})
+    };
+  }
+
+  constructor(adapter = {}) {
+    super(Object.assign(MDLTextfieldFoundation.defaultAdapter, adapter));
+    this.inputFocusHandler_ = () => this.activateFocus_();
+    this.inputBlurHandler_ = () => this.deactivateFocus_();
+  }
+
+  init() {
+    this.adapter_.addClass(MDLTextfieldFoundation.cssClasses.UPGRADED);
+    this.adapter_.registerInputFocusHandler(this.inputFocusHandler_);
+    this.adapter_.registerInputBlurHandler(this.inputBlurHandler_);
+  }
+
+  destroy() {
+    this.adapter_.removeClass(MDLTextfieldFoundation.cssClasses.UPGRADED);
+    this.adapter_.deregisterInputFocusHandler(this.inputFocusHandler_);
+    this.adapter_.deregisterInputBlurHandler(this.inputBlurHandler_);
+  }
+
+  activateFocus_() {
+    const {FOCUSED, LABEL_FLOAT_ABOVE} = MDLTextfieldFoundation.cssClasses;
+    this.adapter_.addClass(FOCUSED);
+    this.adapter_.addClassToLabel(LABEL_FLOAT_ABOVE);
+    this.showHelptext_();
+  }
+
+  showHelptext_() {
+    const {ARIA_HIDDEN} = MDLTextfieldFoundation.strings;
+    this.adapter_.rmHelptextAttr(ARIA_HIDDEN);
+  }
+
+  deactivateFocus_() {
+    const {FOCUSED, INVALID, LABEL_FLOAT_ABOVE} = MDLTextfieldFoundation.cssClasses;
+    const input = this.getNativeInput_();
+    const isValid = input.checkValidity();
+
+    this.adapter_.removeClass(FOCUSED);
+    if (!input.value) {
+      this.adapter_.removeClassFromLabel(LABEL_FLOAT_ABOVE);
+    }
+    if (isValid) {
+      this.adapter_.removeClass(INVALID);
+    } else {
+      this.adapter_.addClass(INVALID);
+    }
+    this.updateHelptextOnDeactivation_(isValid);
+  }
+
+  updateHelptextOnDeactivation_(isValid) {
+    const {HELPTEXT_PERSISTENT, HELPTEXT_VALIDATION_MSG} = MDLTextfieldFoundation.cssClasses;
+    const {ROLE} = MDLTextfieldFoundation.strings;
+    const helptextIsPersistent = this.adapter_.helptextHasClass(HELPTEXT_PERSISTENT);
+    const helptextIsValidationMsg = this.adapter_.helptextHasClass(HELPTEXT_VALIDATION_MSG);
+    const validationMsgNeedsDisplay = helptextIsValidationMsg && !isValid;
+
+    if (validationMsgNeedsDisplay) {
+      this.adapter_.setHelptextAttr(ROLE, 'alert');
+    } else {
+      this.adapter_.rmHelptextAttr(ROLE);
+    }
+
+    if (helptextIsPersistent || validationMsgNeedsDisplay) {
+      return;
+    }
+    this.hideHelptext_();
+  }
+
+  hideHelptext_() {
+    const {ARIA_HIDDEN} = MDLTextfieldFoundation.strings;
+    this.adapter_.setHelptextAttr(ARIA_HIDDEN, 'true');
+  }
+
+  isDisabled() {
+    return this.getNativeInput_().disabled;
+  }
+
+  setDisabled(disabled) {
+    const {DISABLED} = MDLTextfieldFoundation.cssClasses;
+    this.getNativeInput_().disabled = disabled;
+    if (disabled) {
+      this.adapter_.addClass(DISABLED);
+    } else {
+      this.adapter_.removeClass(DISABLED);
+    }
+  }
+
+  getNativeInput_() {
+    return this.adapter_.getNativeInput() || {
```

Comment:
Take Angular2, for example. An `ElementRef` may be undefined until the component is initialized, and it may be further undefined for SSR. Similarly, clients may have removed and/or re-arranged the DOM, for example in a virtual list. We want to make sure we can be resilient to `getNativeInput()` return a type that cannot be dereferenced.